src/rust/engine/fs/src/snapshot.rs

```diff
@@ -148,6 +149,125 @@ impl Snapshot {
       .to_boxed()
   }
 
+  ///
+  /// Given N Snapshots, returns a new Snapshot that merges them.
+  ///
+  /// Any files that exist in multiple Snapshots will cause this method to fail: the assumption
+  /// behind this behaviour is that almost any colliding file would represent a Rule implementation
+  /// error, and in cases where overwriting a file is desirable, explicitly removing a duplicated
+  /// copy should be straightforward.
+  ///
+  pub fn merge(store: Store, snapshots: &[Snapshot]) -> BoxFuture<Snapshot, String> {
+    // We dedupe PathStats by their symbolic names, as those will be their names within the
+    // `Directory` structure. Only `Dir+Dir` collisions are legal.
+    let path_stats = {
+      let mut uniq_paths: OrderMap<PathBuf, PathStat> = OrderMap::new();
+      for path_stat in snapshots
+        .iter()
+        .map(|s| s.path_stats.iter().cloned())
+        .flatten()
+      {
+        match uniq_paths.entry(path_stat.path().to_owned()) {
+          ordermap::Entry::Occupied(e) => match (&path_stat, e.get()) {
+            (&PathStat::Dir { .. }, &PathStat::Dir { .. }) => (),
+            (x, y) => {
+              return future::err(format!(
+                "Snapshots contained duplicate path: {:?} vs {:?}",
+                x, y
+              )).to_boxed()
+            }
+          },
+          ordermap::Entry::Vacant(v) => {
+            v.insert(path_stat);
+          }
+        }
+      }
+      uniq_paths.into_iter().map(|(_, v)| v).collect()
+    };
+    // Recursively merge the Digests in the Snapshots.
+    Self::merge_helper(store, snapshots.iter().map(|s| s.digest).collect())
+      .map(move |root_digest| Snapshot {
+        digest: root_digest,
+        path_stats: path_stats,
+      })
+      .to_boxed()
+  }
+
+  ///
+  /// Given Digest(s) representing Directory instances, merge them recursively into a single
+  /// output Directory Digest. Fails for collisions.
+  ///
+  fn merge_helper(store: Store, dir_digests: Vec<Digest>) -> BoxFuture<Digest, String> {
+    if dir_digests.is_empty() {
+      // TODO: This will not have been stored... we'll need to explicitly store it.
+      return future::ok(EMPTY_DIGEST).to_boxed();
+    } else if dir_digests.len() == 1 {
+      let mut dir_digests = dir_digests;
```

Comment:
Would possibly be a little clearer to make the argument `mut`, rather than to `mut`-ify it in the method body

Reply:
I did this to avoid that... this localizes the awareness of the mutability to where it is actually relevant.